judge/models/tests/test_contest.py

```diff
@@ -0,0 +1,454 @@
+        del contest.get_label_for_problem
+        contest.full_clean()
+
+    def test_normal_user_current_contest(self):
+        current_contest = self.profiles['normal'].current_contest
+        self.assertIsNotNone(current_contest)
+
+        current_contest.set_disqualified(True)
+        self.profiles['normal'].refresh_from_db()
+        self.assertTrue(current_contest.is_disqualified)
+        self.assertIsNone(self.profiles['normal'].current_contest)
+        self.assertEqual(current_contest.score, -9999)
+
+        current_contest.set_disqualified(False)
+        self.profiles['normal'].refresh_from_db()
+        self.assertFalse(current_contest.is_disqualified)
+        self.assertIsNone(self.profiles['normal'].current_contest)
+        self.assertEqual(current_contest.score, 0)
+
+    def test_live_participation(self):
+        participation = ContestParticipation.objects.get(
+            contest=self.hidden_scoreboard_contest,
+            user=self.profiles['normal'],
+            virtual=ContestParticipation.LIVE,
+        )
+        self.assertTrue(participation.live)
+        self.assertFalse(participation.spectate)
+        self.assertEqual(participation.end_time, participation.contest.end_time)
+        self.assertFalse(participation.ended)
+        self.assertIsInstance(participation.time_remaining, timezone.timedelta)
+
+    def test_spectating_participation(self):
+        participation = create_contest_participation(
+            contest='hidden_scoreboard',
+            user='superuser',
+            virtual=ContestParticipation.SPECTATE,
+        )
+
+        self.assertFalse(participation.live)
+        self.assertTrue(participation.spectate)
+        self.assertEqual(participation.start, participation.contest.start_time)
+        self.assertEqual(participation.end_time, participation.contest.end_time)
+
+    def test_virtual_participation(self):
+        participation = create_contest_participation(
+            contest='private',
+            user='superuser',
+            virtual=1,
+        )
+
+        self.assertFalse(participation.live)
+        self.assertFalse(participation.spectate)
+        self.assertEqual(participation.start, participation.real_start)
+        self.assertIsInstance(participation.end_time, timezone.datetime)
+
+
+class ContestTagTestCase(TestCase):
+    @classmethod
+    def setUpTestData(self):
+        self.basic_tag = ContestTag.objects.create(
+            name='basic',
+            color='#fff',
+        )
+        self.dark_tag = ContestTag.objects.create(
+            name='dark',
+            color='#010001',
+        )
+
+    def test_basic_tag(self):
+        self.assertEqual(str(self.basic_tag), self.basic_tag.name)
+        self.assertEqual(self.basic_tag.text_color, '#000')
+
+    def test_dark_tag(self):
+        self.assertEqual(self.dark_tag.text_color, '#fff')
+
+
+class MinValueOrNoneValidatorTestCase(SimpleTestCase):
+    def test_success(self):
```

Comment:
This should probably be split up into individual tests with more verbose names to fall in line with similar test like those for DiggPaginator and InfinitePaginator.

Reply:
Done